imap_processing/swapi/l1/swapi_l1.py

```diff
@@ -1 +1,565 @@
+    diff = sweep[1:] - sweep[:-1]
+
+    # This uses sliding window to find index where cycle starts.
+    # This is what this below code line is doing:
+    # [1 0 0 1 0 0 0 0 0 1 0 0 1 0 0 0 0]      # Is cycle zero?
+    # [1 1 0 1 1 1 0 1 0 0 1 0 1 1 1 0 1]      # Next diff is one?
+    #   [1 0 1 1 1 0 1 0 0 1 0 1 1 1 0 1 0]    # Next diff is one?
+    #     [0 1 1 1 0 1 0 0 1 0 1 1 1 0 1 0 0]  # Next diff is one?
+    #
+    # [0 0 0 1 0 0 0 0 0 0 0 0 1 0 0 0 0]      # And all?
+
+    ione = diff == 1
+
+    valid = (
+        (sweep == 0)[:-11]
+        & ione[:-10]
+        & ione[1:-9]
+        & ione[2:-8]
+        & ione[3:-7]
+        & ione[4:-6]
+        & ione[5:-5]
+        & ione[6:-4]
+        & ione[7:-3]
+        & ione[8:-2]
+        & ione[9:-1]
+        & ione[10:]
+    )
+    return np.where(valid)[0]
+
+
+def get_indices_of_full_sweep(seq_number: np.ndarray):
+    """Get indices of full cycles.
+
+    Beginning of a sweep is marked by SWP_SCI.SEQ_NUMBER=0
+    (Sequence number of set of steps in energy sweep);
+    end of a sweep is marked by SWP_SCI.SEQ_NUMBER=11;
+    all packets must be present to process a sweep.
+
+    In this function, we get the indices of SEQ_NUMBER
+    0 and then construct full sweep indices.
+
+    Parameters
+    ----------
+    seq_number : numpy.ndarray
+        Array that contains SEQ_NUMBER data informations.
+        Eg. sci_dataset["SEQ_NUMBER"].data
+
+    Returns
+    -------
+    numpy.ndarray
+        1D array with indices of full cycle data.
+    """
+    indices_of_start = find_sweep_starts(seq_number)
+    # find_sweep_starts[..., None] creates array of shape(n, 1).
+    #   Eg. [[3], [8]]
+    # np.arange(12)[None, ...] creates array of shape(1, 12)
+    #   Eg. [[0, 1, 2, 3, ....., 11]]
+    # then we add both of them together to get an array of shape(n, 4)
+    #   Eg. [[3, 4, 5, 6,...14], [8, 9, 10, 11, ..., 19]]
+    full_cycles_indices = indices_of_start[..., None] + np.arange(12)[None, ...]
+    return full_cycles_indices.reshape(-1)
+
+
+def filter_full_cycle_data(full_cycle_data_indices: np.ndarray, l1a_data: xr.Dataset):
+    """Filter metadata and science of packets that makes full cycles.
+
+    Parameters
+    ----------
+    full_cycle_data_indices : numpy.ndarray
+        Array with indices of full cycles.
+    l1a_data : xarray.dataset
+        L1A dataset
+
+    Returns
+    -------
+    xarray.dataset
+        L1A dataset with filtered metadata.
+    """
+    # Had to create new xr.Dataset because Epoch shape and new data variables shapes was
+    # different.
+    full_sweep_dataset = xr.Dataset(
+        coords={"Epoch": l1a_data["Epoch"].data[full_cycle_data_indices]}
+    )
+    for key, value in l1a_data.items():
+        full_sweep_dataset[key] = xr.DataArray(value.data[full_cycle_data_indices])
+    return full_sweep_dataset
+
+
+def uncompress_count(count_data: np.ndarray, compress_indicator: np.ndarray = None):
```

Comment:
I _think_ we were going with **de**compress everywhere? @bourque would probably know.
```suggestion
def decompress_count(count_data: np.ndarray, compress_indicator: np.ndarray = None):
```

Reply:
I agree, it should be `decompress` or at least that is what codice uses. This function is decompressing not uncompressing in my view.